Add missing data in streams

laurentS opened the pull request:
This is a duplicate of #68 (now closed) from inside the repo to test CI access to env vars.

This PR updates several streams to use the sdk schema definition format instead of JSON files and:

- projects now includes (optionally) a license object
- issues includes more complete author and assignees objects which were missing
- issues and MRs include notes/comments
- the `languages` stream was added

The tap's output did not necessarily match the data returned by the api, so this might seem to break compatibility, but I think the missing fields were null anyway.

laurentS commented:
@aaronsteers you seem to have done quite a bit of work on this tap recently, and I got the impression that you were testing it with meltano beyond the CI checks. Hopefully I've not broken anything when merging your changes into this branch :crossed_fingers:

laurentS commented:
@aaronsteers I added the last 2 commits because despite passing CI checks, the output of the tap broken when fed into `target-postgres`. I couldn't find any checks in the sdk that would check for the problems I got, so added 2 tests here. I think they might be useful in the sdk itself, if it makes sense to add them there.

aaronsteers commented on tap_gitlab/tests/test_core.py:
```diff
@@ -43,3 +48,28 @@ def test_tap_config_defaults():
         assert f"fetch_{optin_stream}" in tap.config
+
+
+def test_pkeys_in_schema():
+    """Verify that primary keys are actually in the stream's schema."""
+    tap = TapGitLab(config=SAMPLE_CONFIG, parse_env_config=True)
+    for name, stream in tap.streams.items():
+        pkeys = stream.primary_keys or []
+        schema_props = set(stream.schema["properties"].keys())
+        for pkey in pkeys:
+            assert (
+                pkey in schema_props
+            ), f"Coding error in stream {name}: primary_key {pkey} is missing in schema"
```
This was a big recurring bug in my testing. Thanks for adding this test.

laurentS replied:
Would it make sense to add such checks in the sdk itself? I don't know how well it applies to other types of taps, say ones with more dynamic schema, but it seems to me that a tap should not be able to pass tests/CI checks and still return invalid records.

I can open a PR in the `sdk` if you think it makes sense.

laurentS replied:
Closing this as the test PR in the sdk was merged, and I just upgraded the sdk version here.

aaronsteers commented on tap_gitlab/schemas/project_labels.json:
```diff
@@ -1,41 +0,0 @@
-{
```
A few others were deleted as well.

aaronsteers replied:
Nevermind. I missed this due to GitHub auto-collapsing the `streams.py` folder.

aaronsteers reviewed (changes requested):
A few things in here I'm not really understanding as of yet. Comments/questions inline. Thanks!

aaronsteers commented on tap_gitlab/streams.py:
```diff
@@ -54,6 +75,111 @@ def get_child_context(self, record: dict, context: Optional[dict]) -> dict:
+        Projects endpoint can include license info if explicitly requested.
+        """
+        params = super().get_url_params(context, next_page_token)
+        if self.config.get("include_license_info_for_projects", False):
+            params["license"] = True
+        return params
```
Can we default this to `true` and remove the config option?

laurentS replied:
Good idea. Done.

aaronsteers commented on tap_gitlab/schemas/projects.json:
```diff
@@ -1,250 +0,0 @@
-{
```
Nevermind. I missed this due to GitHub auto-collapsing the `streams.py` folder.

aaronsteers commented on archive/tap_gitlab/__init__.py:
```diff
@@ -17,251 +17,261 @@
-    'fetch_pipelines_extended': False,
-    'fetch_group_variables': False,
-    'fetch_project_variables': False,
+    "api_url": "https://gitlab.com/api/v4",
```
I would not mind if you just want to delete the `archive` folder in this PR. That would actually reduce the lines of code needing to be reviewed here. We can always compare with the legacy branch if we want to in the future.

laurentS replied:
I made no changes indeed, but you're right that the autoformat caught this in its net. sorry about this.
I've now removed the `archive` folder, but moved over the `LICENSE` file to keep it.

laurentS commented on tap_gitlab/client.py:
```diff
@@ -144,13 +145,53 @@ def post_process(self, row: dict, context: Optional[dict] = None) -> Optional[di
     """Base class for streams that are keys based on group ID."""
 
-    state_partitioning_keys = ["group_path"]
+    state_partitioning_keys = ["group_id"]
```
My reasoning was strongly influenced by what we've seen in [this tap-github issue](https://github.com/MeltanoLabs/tap-github/issues/110).
Paths can be renamed, potentially leading to inconsistencies, whereas `id`s should remain stable across renames. Depending on how the tap is used (which `group` values are passed in config) passing `newOrgName` in config while `oldOrgName` is in the state from a previous run would cause the tap to miss the bookmarks entirely, and potentially redownload a lot of data. Using `group_id` should prevent such problems.

The result is indeed a bit less readable than using the `group_path`.

Happy to hear your thoughts on this.

aaronsteers replied:
@laurentS - I had not thought about renames! Very good argument - thank you. I'm inclined to agree with this change, although I think it should be considered a breaking change.

I'll open a new issue to consolidate and strategize around breaking changes like this one.

laurentS replied:
See my note below. I just added this, as I think it makes sense. The tap should now run on numeric ids instead of paths, both for `projects` and `groups`. Casing is corrected and invalid names are removed from the list of partitions.

laurentS commented on tap_gitlab/client.py:
```diff
@@ -170,5 +211,5 @@ def partitions(self) -> List[dict]:
             "Could not detect partition type for Gitlab stream "
             f"'{self.name}' ({self.path}). "
-            "Expected a URL path containing '{project_path}' or '{group_path}'. "
+            "Expected a URL path containing '{project_path}' or '{group_id}'. "
```
This relates to the comment above. I'd be tempted to do something similar to https://github.com/MeltanoLabs/tap-github/pull/112/files to ensure we have correct group/project ids/paths for everything before the tap really runs. If you think that's a valid option, then I'll implement here as well. Waiting for your feedback before I get into this.

laurentS replied:
@aaronsteers how do you suggest handling the breaking change? do you have a clear policy for this at meltano?

laurentS replied:
@aaronsteers Correct, the behaviour you describe is what the code does. Config inputs expect strings (eg. `meltano/sdk`) but the first thing the tap will do is cleanup these inputs (for casing, non-existing repos...) and then use numeric ids internally wherever possible. Some urls will use paths where IDs are not supported, but state and pkeys are all numerics.

laurentS replied:
@aaronsteers Correct, the behaviour you describe is what the code does. Config inputs expect strings (eg. `meltano/sdk`) but the first thing the tap will do is cleanup these inputs (for casing, non-existing repos...) and then use numeric ids internally wherever possible. Some urls will use paths where IDs are not supported, but state and pkeys are all numerics.

laurentS commented:
At this point, I think this PR is ready. We have been running the code in a staging env for a while, with no major problems.
I think a breaking change note would help, something like:
> This release introduces breaking changes in the state management of the tap: streams now use numeric ids to track projects and groups to better handle changes in names. If you have bookmarks from a previous version, they will become invalid, meaning the tap will restart streams "from the beginning" the first time you run on this release.

edgarrmondragon reviewed (approved):
@laurentS @ericboucher this looks good to me, though the breaking change makes me want to merge this to a preview branch.

A warning printed to stderr would be nice, additionally.

Wdyt @aaronsteers?

laurentS commented:
> @laurentS @ericboucher this looks good to me, though the breaking change makes me want to merge this to a preview branch.

 
> A warning printed to stderr would be nice, additionally.

How about a warning in `TapGitlab.__init__()` ? Something like "This version of the tap contains a number of breaking changes (streams, schema, state...). Please read the changelog for details"?
 
> Wdyt @aaronsteers?

laurentS commented:
@aaronsteers @edgarrmondragon This PR has been hanging for a while, sorry. I've updated the `sdk` to the latest version, updated the tests that have now moved in there, added a note in the README, and a runtime warning (in `load_state`) about the breaking change in the state format.
I've also updated a bunch of development related packages, `poetry`, `mypy`, `black`, `flake8` and the github actions used here.`
In my mind, this is ready to merge. :rocket: Over to you :)

aaronsteers reviewed (approved):
I'm approving and I've also bumped version to `2.0`.

Notes:

- Default branch as of now is still `legacy-stable`. This is because legacy references to this tap did not use any type of version pinning.
- The `main` branch is not default as of now, but users can reference it explicitly or (better) use the new `v2.0.0` release tag I'm cutting now.
